### PR TITLE
Dynamic configuration API and S3 support

### PR DESCRIPTION
**Pull request by andrebsguedes**
Conforms to the new dynamic config API in `object_store_ffi_jl` `v0.2.0` and expands support to AWS S3.

The main functions names and signatures from the public API are also changed (`blob_get! => get!`, `blob_put => put` ...)

### CODE REVIEW

File: src/RustyObjectStore.jl [hunk 1]
```diff
@@ -30,20 +36,31 @@ Global configuration for the object store requests.
...
-    max_retries::Culonglong
-    "The number of seconds from the initial request after which no further retries will be attempted."
-    retry_timeout_sec::Culonglong
+@kwdef struct StaticConfig
+    "The number of worker threads for the native pool (0 == auto)"
+    n_threads::Culonglong
+    "The maximum capacity for the client cache"
+    cache_capacity::Culonglong
+    "The time-to-live in seconds for entries in the client cache"
+    cache_ttl_secs::Culonglong
```
Comment (jfunstonRAI) [inline src/RustyObjectStore.jl:45]: I'm assuming 0 here will disable ttl and not have it instantly expired?

File: src/RustyObjectStore.jl [hunk 2]
```diff
@@ -81,52 +98,193 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+- `request_timeout_secs::Int`: (Optional) Client request timeout in seconds.
+- `connect_timeout_secs::Int`: (Optional) Client connection timeout in seconds.
+- `max_retries::Int`: (Optional) Maximum number of retry attempts.
+- `retry_timeout_secs::Int`: (Optional) Maximum amount of time from the initial request after which no further retries will be attempted (in seconds).
+"""
+struct ClientOptions
+    params::Dict{String, String}
+
+    function ClientOptions(;
+        request_timeout_secs::Int = 30,
```
Comment (jfunstonRAI) [inline src/RustyObjectStore.jl:114]: Can we make these options and use the default on the rust side if they aren't set?

File: test/aws_s3_tests.jl [hunk 3]
```diff
@@ -0,0 +1,225 @@
+@testitem "Basic AWS S3 usage" setup=[InitializeObjectStore] begin
+using CloudBase.CloudTest: Minio
+using RustyObjectStore: get!, put, AwsConfig, ClientOptions
+
+# For interactive testing, use Minio.run() instead of Minio.with()
+# conf, p = Minio.run(; debug=true, public=false); atexit(() -> kill(p))
+Minio.with(; debug=true, public=false) do conf
```
Comment (jfunstonRAI) [inline test/aws_s3_tests.jl:7]: Can we parameterize these tests so that they don't have to be duplicated? Maybe like how we have it in the pager where the actual test is in a function and then we call that inside Minio and Azurite blocks.

File: src/RustyObjectStore.jl (hunk 1 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:45]: Well, I would have hoped too, but the way I did it on the Rust side will actually make it expire instantly. Sadly changing this requires a new release of object_store_ffi_jl. We could leave that as an issue on the Rust side so we remember to make this change for the next release. Do you think I should place a guard here, in the mean time, to set a huge default duration in the case zero is provided?

File: src/RustyObjectStore.jl (hunk 2 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:114]: Will do!

File: test/aws_s3_tests.jl (hunk 3 above)
Comment (andrebsguedes) [reply test/aws_s3_tests.jl:7]: I dislike the duplication here too, in the specific case of these tests, I think we should be good with a single implementation.

But there are other tests just as duplicated as these that actually have different errors for different store backends which makes it a little cumbersome to make them generic.

Do you think it is worth it to pursue removing all duplication or just part of it is enough?

File: src/RustyObjectStore.jl (hunk 1 above)
Comment (jfunstonRAI) [reply src/RustyObjectStore.jl:45]: So the fix on the rust side is if we are passed 0 then we just don't set the config option on the cache, right? I can include that on my PR with the multipart operations, just so we don't keep bouncing around with PRs for version bumps. In that case I don't think you need to change anything here.

File: test/aws_s3_tests.jl (hunk 3 above)
Comment (jfunstonRAI) [reply test/aws_s3_tests.jl:7]: I think just the easy to dedup happy path tests are enough for now.

File: src/RustyObjectStore.jl (hunk 1 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:45]: Yes, exactly like the `n_threads` config in the runtime. Thanks!

File: src/RustyObjectStore.jl [hunk 4]
```diff
@@ -81,52 +98,193 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+            first = false
+        else
+            print(io, ", ")
+        end
+        if occursin(r"secret|token|key", string(key))
+            print(io, repr(key), " => ", repr("*****"))
+        else
+            print(io, repr(key), " => ", repr(dict[key]))
+        end
+    end
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:278]: This `show` message doesn't round-trip, which is unfortunate.

it's very helpful to have `show` messages that display objects as valid Julia code for reconstructing an object.

in general, the string representation should parse to valid Julia code which when evaluated produces something equal to the original object, i.e. we should have `x == eval(Meta.parse(repr(x)))`

(potentially with additional "pretty-printing" method, defined by a 3-argument `show` method, but i don't think we need those here)

In practice this is useful for debugging... if you have some object `az`, and want to create a similar object (but don't know how), the natural thing to do is copy-paste its string representation
```julia
julia> az
AzureConfig(Config(:max_retries => "10", :url => "az://my_container/", :retry_timeout_secs => "150", :timeout => "30s", :azure_container_name => "my_container", :connect_timeout => "5s", :azure_storage_account_name => "my_account"))

julia> AzureConfig(Config(:max_retries => "10", :url => "az://my_container/", :retry_timeout_secs => "150", :timeout => "30s", :azure_container_name => "my_container", :connect_timeout => "5s", :azure_storage_account_name => "my_account"))
ERROR: MethodError: no method matching Config(::Pair{Symbol, String}, ::Pair{Symbol, String}, ::Pair{Symbol, String}, ::Pair{Symbol, String}, ::Pair{Symbol, String}, ::Pair{Symbol, String}, ::Pair{Symbol, String})
```

File: src/RustyObjectStore.jl [hunk 5]
```diff
@@ -1,11 +1,17 @@
 module RustyObjectStore
 
-export init_object_store, blob_get!, blob_put, AzureCredentials, ObjectStoreConfig
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:3]: Are we purposefully not exporting `StaticConfig` or was that just an oversight?

File: src/RustyObjectStore.jl [hunk 6]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+        max_retries::Option{Int} = nothing,
+        retry_timeout_secs::Option{Int} = nothing
+    )
+        params = Dict()
+        if !isnothing(request_timeout_secs)
+            params["timeout"] = "$(string(request_timeout_secs))s"
+        end
+
+        if !isnothing(connect_timeout_secs)
+            params["connect_timeout"] = "$(string(connect_timeout_secs))s"
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:125]: why do `connect_timeout_secs` and `request_timeout_secs` need a trialing `s`, but `retry_timeout_secs` doesn't?

File: src/RustyObjectStore.jl [hunk 7]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+
+    function ClientOptions(;
+        request_timeout_secs::Option{Int} = nothing,
+        connect_timeout_secs::Option{Int} = nothing,
+        max_retries::Option{Int} = nothing,
+        retry_timeout_secs::Option{Int} = nothing
+    )
+        params = Dict()
+        if !isnothing(request_timeout_secs)
+            params["timeout"] = "$(string(request_timeout_secs))s"
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:121]: more idiomatic than calling `string` inside string interpolation would be to just use `string` with multiple arguments, like:
```suggestion
            params["timeout"] = string(request_timeout_secs, "s")
```
(same for the other similar cases)

File: src/RustyObjectStore.jl [hunk 8]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+
+function as_config(wrapper::AsConfig) end
+
 """
     $TYPEDEF
 
 # Arguments
-$TYPEDFIELDS
+- `url::String`: Url of the object store container root path.
+  It must include the cloud specific url scheme (s3://, azure:// ...).
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:149]: is there a known fixed list of url schemes we support? (is it just `s3://`, `az://`, `azure://`?) 

if so, can we just list them all here?

File: src/RustyObjectStore.jl [hunk 9]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+            params["retry_timeout_secs"] = string(retry_timeout_secs)
+        end
+
+        return new(params)
+    end
+end
+
+abstract type AsConfig end
+
+function as_config(wrapper::AsConfig) end
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:142]: what is this trying to do?

This just defines a new method that returns `nothing`, is that what's intended?

File: src/RustyObjectStore.jl [hunk 10]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+
+        if !isnothing(retry_timeout_secs)
+            params["retry_timeout_secs"] = string(retry_timeout_secs)
+        end
+
+        return new(params)
+    end
+end
+
+abstract type AsConfig end
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:140]: why is this named `AsConfig`? please can we give this a more informative name?

is this just providing a common supertype for the AWS and Azure configs? If so, maybe just `AbstractConfig` would be fine?

File: src/RustyObjectStore.jl [hunk 11]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
-    account::String
-    "Azure container name"
-    container::String
-    "Azure access key"
-    key::String
-    "(Optional) Alternative Azure host. For example, if using Azurite."
-    host::String
-    function AzureCredentials(account::String, container::String, key::String, host::String="")
-        return new(account, container, key, host)
+struct Config <: AsConfig
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:153]: why does this have the same supertype as the `AzureConfig` and `AWSConfig`? 

I presumed this was just an internal helper, and not user-facing, so a user wouldn't ever actually constuct this and pass it around like they would AWS/AzureConfigs -- but perhaps i'm misunderstanding?

**Review by jfunstonRAI — APPROVED** [review]

File: src/RustyObjectStore.jl [hunk 12]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+# Keyword Arguments
+- `storage_account_name::String`: Azure storage account name.
+- `container_name::String`: Azure container name.
+- `storage_account_key::Option{String}`: (Optional) Azure storage account key.
+- `storage_sas_token::Option{String}`: (Optional) Azure storage SAS token.
+- `host::Option{String}`: (Optional) Alternative Azure host. For example, if using Azurite.
+- `opts::ClientOptions`: (Optional) Client configuration options.
+"""
+struct AzureConfig <: AsConfig
+    config::Config
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:176]: Why store this as `Config` rather than just a `String`? Does it make FFI cleaner somehow to guarantee they both go through this type? 

I wonder if `Config` could instead just be a function, `config_string(params...)`, that calls `JSON3.write`?

File: src/RustyObjectStore.jl [hunk 13]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+        container_name::String,
+        storage_account_key::Option{String} = nothing,
+        storage_sas_token::Option{String} = nothing,
+        host::Option{String} = nothing,
+        opts::ClientOptions = ClientOptions()
+    )
+        (
+            !isnothing(storage_account_key)
+            && !isnothing(storage_sas_token)
+        ) && error("Should provide either a storage_account_key or a storage_sas_token")
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:188]:
```suggestion
        if !isnothing(storage_account_key) && !isnothing(storage_sas_token)
            error("Should provide either a storage_account_key or a storage_sas_token"
        end
```

File: src/RustyObjectStore.jl [hunk 14]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
-    print(io, repr(credentials.account), )
-    print(io, ", ", repr(credentials.container))
-    print(io, ", ", "\"*****\"") # don't print the secret key
-    !isempty(credentials.host) && print(io, ", ", repr(credentials.host))
+
+as_config(wrapper::Config) = wrapper
+
+"""
+    $TYPEDEF
+
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:166]: docstring should mention the need to provide exactly one of `storage_account_key` or `storage_sas_token`

File: src/RustyObjectStore.jl [hunk 15]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+        params["azure_storage_account_name"] = storage_account_name
+        params["azure_container_name"] = container_name
+
+        if !isnothing(storage_account_key)
+            params["azure_storage_account_key"] = storage_account_key
+        end
+
+        if !isnothing(storage_sas_token)
+            params["azure_storage_sas_token"] = storage_sas_token
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:200]: is it fine to pass _both_ a `storage_account_key` and a `storage_sas_token`?

File: src/RustyObjectStore.jl [hunk 16]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+            print(io, repr(key), " => ", repr("*****"))
+        else
+            print(io, repr(key), " => ", repr(dict[key]))
+        end
+    end
     print(io, ")")
 end
 
-const _AzureCredentialsFFI = NTuple{4,Cstring}
+const _ConfigFFI = NTuple{1,Cstring}
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:282]: Why do we need this to be a `Tuple{CString}` rather than just a `CString`?

File: src/RustyObjectStore.jl [hunk 17]
```diff
@@ -81,52 +98,203 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+
+        if !isnothing(session_token)
+            params["aws_session_token"] = session_token
+        end
+
+        if !isnothing(host)
+            params["minio_host"] = host
+        end
+
+        return new(Config("s3://$(bucket_name)/", params))
```
Comment (nickrobinson251) [inline src/RustyObjectStore.jl:257]: why do we "eagerly" turn all the given params into a single `String`, rather than storing the data as given and then just turning it into a string when needed for a `get!` or `put` call

I presume for efficiency (i.e. so we only need to create each config string once, no matter how many `get!`/`put` calls)

Can we add that to a comment somewhere? If that's an important consideration, i think we should document it so it never gets refactored away

**Review by nickrobinson251 — COMMENTED** [review]
Left some comments, largely stylistic. Not reviewed `builder` or `test/` yet.

File: src/RustyObjectStore.jl [hunk 18]
```diff
@@ -1,11 +1,17 @@
 module RustyObjectStore
 
-export init_object_store, blob_get!, blob_put, AzureCredentials, ObjectStoreConfig
+export init_object_store, get!, put, ClientOptions, Config, AzureConfig, AwsConfig
```
[the remark this answers is not in the document]
Comment (jfunstonRAI) [reply src/RustyObjectStore.jl:3]: In that case should we also not export `put` for consistency?

File: src/RustyObjectStore.jl (hunk 4 above)
Comment (jfunstonRAI) [reply src/RustyObjectStore.jl:278]: How does that work with the masked secrets? Would we only have that for the pretty-print method? If we log something does it use show or pretty-print?

File: src/RustyObjectStore.jl [hunk 19]
```diff
@@ -30,20 +36,31 @@ Global configuration for the object store requests.
...
 $TYPEDFIELDS
 """
-@kwdef struct ObjectStoreConfig
-    "The maximum number of times to retry a request."
-    max_retries::Culonglong
-    "The number of seconds from the initial request after which no further retries will be attempted."
-    retry_timeout_sec::Culonglong
+@kwdef struct StaticConfig
+    "The number of worker threads for the native pool (0 == auto)"
+    n_threads::Culonglong
```
[the remark this answers is not in the document]
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:41]: Automatic detection, which means `n_threads == num_cpus`. Will clarify!

File: src/RustyObjectStore.jl (hunk 18 above)
[the remark this answers is not in the document]
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:3]: Will omit `get!` from the export, if changing the name is preferred please make a suggestion.

File: src/RustyObjectStore.jl [hunk 20]
```diff
@@ -81,52 +98,193 @@ function init_object_store(config::ObjectStoreConfig=DEFAULT_CONFIG)
...
+# Keyword Arguments
+- `region::String`: AWS S3 region.
+- `bucket_name::String`: AWS S3 bucket name.
+- `access_key_id::Option{String}`: (Optional) AWS S3 access key id.
+- `secret_access_key::Option{String}`: (Optional) AWS S3 secret access key.
+- `session_token::Option{String}`: (Optional) AWS S3 session_token.
+- `host::Option{String}`: (Optional) Alternative S3 host. For example, if using Minio.
+- `opts::ClientOptions`: (Optional) Client configuration options.
+"""
+struct AwsConfig <: AsConfig
```
[the remark this answers is not in the document]
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:225]: Sure!

File: src/RustyObjectStore.jl (hunk 5 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:3]: Oversight. Will fix!

File: src/RustyObjectStore.jl (hunk 6 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:125]: `request_timeout_secs` and `connect_timeout_secs` are parsed by the object_store library directly into duration objects and it expects something like `30s` meaning 30 seconds.

The `retry_timeout_secs` is parsed by us on the Rust side (because we are making up for a parameter that is not exposed in `object_store`'s API) and we assume the input will always be just the number of seconds.

It would be good to use the same parsing as the other ones but the parsing function is private and I didn't think it warranted an implementation (maybe a crate already exists, I didn't check).

Do you think we should do anything differently here?

File: src/RustyObjectStore.jl (hunk 7 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:121]: Will do!

File: src/RustyObjectStore.jl (hunk 8 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:149]: Sure!

File: src/RustyObjectStore.jl (hunk 10 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:140]: I leaned towards a Rust name I guess, will change to `AbstractConfig` as that sounds more julian

File: src/RustyObjectStore.jl (hunk 6 above)
Comment (jfunstonRAI) [reply src/RustyObjectStore.jl:125]: My vote would be to not complicate the parsing on the rust side if we don't have to.

File: src/RustyObjectStore.jl (hunk 11 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:153]: A user may want to construct this type if he is able to fetch the `url` and `params` directly from a config file and wants to delegate all dispatching and validation to the library dynamically instead of coercing them into one of `AzureConfig` or `AwsConfig`.

File: src/RustyObjectStore.jl (hunk 4 above)
Comment (nickrobinson251) [reply src/RustyObjectStore.jl:278]: unless you define a specific `show(::IO, ::MIME"text/plain", x::MyObject)` method, then it falls back to `show(::IO, x::MyObject)` 

with secrets, we can just return the string "***" or whatever, which means we won't have equality, but we should nevertheless be able to construct a similar object from the show method i.e. show should be a valid constructor (ideally).

i do tend to find this property helpful in debugging, which is why i mention it (rather than it just being aesthetic preferences)

File: src/RustyObjectStore.jl (hunk 12 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:176]: I was assuming that we would want to keep the `cconvert` and `unsafe_convert` implementations limited to a single concrete struct. But thinking about it, we could make those functions operate on the supertype directly and then make do with a string field directly on each of the config types (`Config`, `AzureConfig`, `AWSConfig`). For efficiency we don't want to keep serializing the JSON on each native call, that is why we need a string field on a struct.

Would that be better than wrapping the `Config` type

File: src/RustyObjectStore.jl (hunk 13 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:188]: Will do!

File: src/RustyObjectStore.jl (hunk 14 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:166]: Will do!

File: src/RustyObjectStore.jl (hunk 15 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:200]: Will fix!

File: src/RustyObjectStore.jl (hunk 16 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:282]: Will change!

File: src/RustyObjectStore.jl (hunk 17 above)
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:257]: Yes, it is still debatable if this will end up being significant in terms of performance (mainly in the future). But currently we are doing this for efficiency as you correctly presumed.

With the string stored on the struct we just need to send its pointer to Rust and hash its contents there to find the correct client for the request.

Storing the data as given would be much nicer for sure, but we would need to pay for serialization and allocation for each request (for 2MB requests it certainly does not matter much, but who knows if we may need to do smaller ones in the future).

File: src/RustyObjectStore.jl [hunk 21]
```diff
@@ -1,11 +1,17 @@
 module RustyObjectStore
```
[the remark this answers is not in the document]
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:1]: Will do!

File: src/RustyObjectStore.jl (hunk 17 above)
Comment (nickrobinson251) [reply src/RustyObjectStore.jl:257]: yeah, makes sense -- just worth adding a comment about this somewhere i think

File: src/RustyObjectStore.jl (hunk 6 above)
Comment (nickrobinson251) [reply src/RustyObjectStore.jl:125]: Okay, makes sense... i think i'd add a comment saying why we do this

maybe on the two with the trailing `s`, something like
```
# Include `s` so Rust `Duration` knows to parse this as seconds
```
then on the other one
```
# Parsed directly by `object_store_ffi` as seconds, so don't need to include trailing `s`
```
or something along those lines 

otherwise it seems a bit surprising and inconsistent

File: src/RustyObjectStore.jl (hunk 18 above)
[the remark this answers is not in the document]
Comment (nickrobinson251) [reply src/RustyObjectStore.jl:3]: > if changing the name is preferred please make a suggestion

i did wonder if another name would be simplest, but alas i've not got a good suggestion 🤷 

we can always add an alias in future if we find `RustyObjectStore.get!` too annoying

File: src/builder.jl
Comment (Drvi) [inline src/builder.jl:1]: Just my 2 cents, since others have reviewed in sufficient depth already :+1:: 
* I'm not sure why we added JSON serialization to the mix at the FFI boundary, was that done to ease the development process?
* Locally I played with a variant that uses `Base.ImmutableDict` instead of a `Dict` and did reduce the number of allocations and was overall faster. I mention this because it seems like we care about the performance of constructing these (I guess if someone needs to touch a lot of files and each has a different SAS token, we might end up creating a bunch o these, not sure how realistic this scenario is). Happy to share the code if you think its worth investigating.

File: src/RustyObjectStore.jl (hunk 7 above)
Comment (Drvi) [reply src/RustyObjectStore.jl:121]: Note that the `string` function has a fast path when inputs share the same type, e.g.:
```
julia> @btime string(string(1), "s")
  27.261 ns (3 allocations: 112 bytes)
"1s"

julia> @btime string(1, "s")
  106.952 ns (5 allocations: 240 bytes)
"1s"
```

Comment (nickrobinson251) [reply src/RustyObjectStore.jl:121]: well that's extremely silly. Why wouldn't Julia just dispatch `string(x, y)` to `string(string(x), string(y))` if that's more performant? shall we go do that in Base?

File: src/builder.jl
Comment (nickrobinson251) [reply src/builder.jl:1]: also, if we see Config construction performance as critical, we should add tests/benchmarks for that

Comment (andrebsguedes) [reply src/builder.jl:1]: Actually, it is the opposite haha. Right now we don't care about the performance of constructing the configs exactly because the preferred usage pattern is to construct it only once and reuse it multiple times. That is why we can use JSON instead of manually creating structs that both sides need to safely read and write to.

If constructing the config was very cheap, then we could just store the fields as is and perform the serialization every time, or go back to converting each field through the FFI boundary and dealing with it.

TLDR: Currently we don't care about construction speed exactly because we reuse the result of the construction (the string). We would care about it if we decide that having the serialized string is too ugly.

In hindsight maybe we could just store both the fields and the string (but that can cause its own problems).

File: src/builder.jl [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+        return new(params)
+    end
+end
+
+add_entry(b::ConfigBuilder, key::String, value::String) = b.params[key] = value
+add_entry(b::AzureConfigBuilder, key::String, value::String) =b.params[key] = value
+add_entry(b::AwsConfigBuilder, key::String, value::String) =b.params[key] = value
+
+
+function with_request_timeout_secs(timeout_secs::Int)
```
[the remark this answers is not in the document]
Comment (andrebsguedes) [reply src/builder.jl:61]: I am in favor of ditching the builder altogether, I think the other constructors will do the job just fine

File: src/builder.jl [hunk 23]
```diff
@@ -0,0 +1,172 @@
+abstract type Builder end
+
+function add_entry(b::Builder, key::String, value::String) end
+
+"""
+    $TYPEDEF
+
+Usage samples:
+
+azure_config = ConfigBuilder() |>
```
[the remark this answers is not in the document]
Comment (andrebsguedes) [reply src/builder.jl:10]: As above, I think it is best to let the builder go

File: test/basic_unified_tests.jl [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+        host=base_url
+    )
+
+    run_read_write_test_cases(config)
+
+end # Minio.with
+end # @testitem
+
+@testitem "Basic AWS S3 usage (anonymous read enabled)" setup=[InitializeObjectStore, ReadWriteCases] begin
+# TODO: currently object_store defaults to Instance credentials when no other credentials are supplied
```
[the remark this answers is not in the document]
Comment (andrebsguedes) [reply test/basic_unified_tests.jl:171]: We would need some extra work to prevent object_store from trying to fetch the AWS instance credentials if no other credentials are provided. This would allow anonymous reads but I deemed it not worthy of our time currently as it is unlikely that we are going to use this in the near term. What do you think?

[the remark this answers is not in the document]
Comment (nickrobinson251) [reply test/basic_unified_tests.jl:171]: ah, yeah that decision makes complete sense to me! i just didn't understand that from the comment, maybe let's copy-paste your explanation into an issue and link it here?

File: src/builder.jl
Comment (Drvi) [reply src/builder.jl:1]: Fair enough, thanks!:)

**Review by nickrobinson251 — APPROVED** [review]
feel free to ping me if you want me to look again at anything, otherwise i'll approve based on what we discussed earlier 👍

File: src/RustyObjectStore.jl [hunk 25]
```diff
@@ -1,11 +1,17 @@
...
 
-export init_object_store, blob_get!, blob_put, AzureCredentials, ObjectStoreConfig
+export init_object_store, get_object!, put_object, StaticConfig, ClientOptions, Config, AzureConfig, AWSConfig
+export ConfigBuilder, with_request_timeout_secs, with_connect_timeout_secs, with_max_retries, with_retry_timeout_secs,
+       azure, with_container_name, with_storage_account_name, with_storage_account_key, with_storage_sas_token, build,
```
[the remark this answers is not in the document]
Comment (andrebsguedes) [reply src/RustyObjectStore.jl:5]: Oops
